aten/src/ATen/SparseCsrTensorImpl.cpp

```diff
@@ -0,0 +1,153 @@
+#include <ATen/ATen.h>
+#include <ATen/InitialTensorOptions.h>
+#include <ATen/SparseCsrTensorImpl.h>
+#include <ATen/SparseTensorImpl.h>
+#include <ATen/SparseTensorUtils.h>
+#include <ATen/core/LegacyTypeDispatch.h>
+
+namespace at {
+namespace {
+DeviceType SparseCsrTensorSetToDeviceType(DispatchKeySet key_set) {
+  if (key_set.has(DispatchKey::SparseCsrCPU)) {
+    return kCPU;
+  } else if (key_set.has(DispatchKey::SparseCsrCUDA)) {
+    return kCUDA;
+  } else {
+    AT_ERROR(
+        "Cannot construct SparseTensor with non-sparse tensor type ID ",
+        key_set);
+  }
+}
+} // namespace
+
+SparseCsrTensorImpl::SparseCsrTensorImpl(
+    at::DispatchKeySet key_set,
+    const caffe2::TypeMeta& data_type)
+    : SparseCsrTensorImpl(
+          key_set,
+          data_type,
+          at::empty(
+              {0},
+              at::initialTensorOptions()
+                  .device(SparseCsrTensorSetToDeviceType(key_set))
+                  .dtype(ScalarType::Int)) // crow_indices
+          ,
+          at::empty(
+              {0},
+              at::initialTensorOptions()
+                  .device(SparseCsrTensorSetToDeviceType(key_set))
+                  .dtype(ScalarType::Int)) // col_indices
+          ,
+          at::empty(
+              {0},
+              at::initialTensorOptions()
+                  .device(SparseCsrTensorSetToDeviceType(key_set))
+                  .dtype(data_type)) // values
+      ) {}
+
+SparseCsrTensorImpl::SparseCsrTensorImpl(
+    at::DispatchKeySet key_set,
+    const caffe2::TypeMeta& data_type,
+    at::Tensor crow_indices,
+    at::Tensor col_indices,
+    at::Tensor values)
+    : TensorImpl(key_set, data_type, values.device()),
+      crow_indices_(std::move(crow_indices)),
+      col_indices_(std::move(col_indices)),
+      values_(std::move(values)) {}
+
+void SparseCsrTensorImpl::resize_and_clear_(
+    int64_t nnz_size,
+    IntArrayRef size) {
+  // call crow_indices().options() here since the struct contructor calls the
+  // tensor constructor with args for device specific init.
+  auto empty_crow_indices = at::empty(size[0] + 1, crow_indices().options());
+  auto empty_col_indices = at::empty(nnz_size, col_indices().options());
+  auto empty_values = at::empty(nnz_size, values().options());
+
+  crow_indices_ = empty_crow_indices;
+  col_indices_ = empty_col_indices;
+  values_ = empty_values;
```

Comment:
Why not just resize the internal tensors directly? Saves you from temporarily doubling your memory occ if you resized smaller, and saves you a TensorImpl allocation in all cases

Reply:
Also maintains object identity of indices/values; I don't know how much this matters yet but it can be a useful invariant